src/Analysis/Engine/Impl/Indexing/SymbolIndexWalker.cs

```diff
@@ -0,0 +1,267 @@
+                node.NameExpression.GetSpan(_ast),
+                children,
+                FunctionKind.Class
+            ));
+
+            return false;
+        }
+
+        public override bool Walk(FunctionDefinition node) {
+            _stack.Enter(SymbolKind.Function);
+            foreach (var p in node.Parameters) {
+                AddVarSymbol(p.NameExpression);
+            }
+            node.Body?.Walk(this);
+            var children = _stack.Exit();
+
+            var span = node.GetSpan(_ast);
+
+            var ds = new HierarchicalSymbol(
+                node.Name,
+                SymbolKind.Function,
+                span,
+                node.IsLambda ? span : node.NameExpression.GetSpan(_ast),
+                children,
+                FunctionKind.Function
+            );
+
+            if (_stack.Parent == SymbolKind.Class) {
+                switch (ds.Name) {
+                    case "__init__":
+                        ds.Kind = SymbolKind.Constructor;
+                        break;
+                    case var name when DoubleUnderscore.IsMatch(name):
+                        ds.Kind = SymbolKind.Operator;
+                        break;
+                    default:
+                        ds.Kind = SymbolKind.Method;
+
+                        if (node.Decorators != null) {
+                            foreach (var dec in node.Decorators.Decorators) {
+                                var maybeKind = DecoratorExpressionToKind(dec);
+                                if (maybeKind.HasValue) {
+                                    ds.Kind = maybeKind.Value.Item1;
+                                    ds._functionKind = maybeKind.Value.Item2;
+                                    break;
+                                }
+                            }
+                        }
+
+                        break;
+                }
+            }
+
+            _stack.AddSymbol(ds);
+
+            return false;
+        }
+
+        public override bool Walk(ImportStatement node) {
+            foreach (var imp in node.Names.Zip(node.AsNames, (name, asName) => (name, asName))) {
+                if (imp.asName == null) {
+                    var span = imp.name.GetSpan(_ast);
+                    _stack.AddSymbol(new HierarchicalSymbol(imp.name.MakeString(), SymbolKind.Module, span));
+                } else {
+                    var span = imp.asName.GetSpan(_ast);
+                    _stack.AddSymbol(new HierarchicalSymbol(imp.asName.Name, SymbolKind.Module, span));
+                }
+            }
+
+            return false;
+        }
+
+        public override bool Walk(FromImportStatement node) {
+            foreach (var imp in node.Names.Zip(node.AsNames, (name, asName) => (name, asName))) {
+                if (imp.asName == null) {
+                    var span = imp.name.GetSpan(_ast);
+                    _stack.AddSymbol(new HierarchicalSymbol(imp.name.Name, SymbolKind.Module, span));
```

Comment:
Btw, why don't you want to save reference to `node` in `HierarchicalSymbol`?

Reply:
What do you mean? Hierarchical symbols only refer to their children, and I'm only treating "declared names" as symbols.

Reply:
I think current analysis tends not to keep ASTs around for memory consumption reasons.

Reply:
AST nodes are currently referenced in analysis.

Reply:
Regardless of memory usage, I'm not sure why I'd want to do this. Plus, the symbol classes are technically part of the "public API", so I'm not sure it should need to know anything about the AST.

Reply:
Ok, what do you expect to be the typical usage of `HierarchicalSymbol`?

Reply:
The main use (currently) is to return back as `DocumentSymbol`s. There may be some use past that for finding references, but this walker is pretty simple compared to the real analysis' information (which should be preferred).

Reply:
Ok, let's leave it as it is for now.